File: data/members/insert_records.py
import json
import os
import mysql.connector
from mysql.connector.connection import MySQLConnection
from mysql.connector.cursor import MySQLCursor
from typing import Any, Dict, List
from schema import Record


CONFIG_PATH: str = os.path.join(
    os.path.dirname(
        os.path.dirname(
            os.path.dirname(
                os.path.abspath(
                    __file__)))),
    "config.json"
)
TABLE_NAME: str = "Members"


def get_config() -> Dict[str, Any]:
    with open(CONFIG_PATH, "r", encoding="utf-8") as f:
        return json.load(f)
    

def connect_to_db(config: Dict[str, Any]) -> MySQLConnection:
    return mysql.connector.connect(
        host=config["SQL_HOST_NAME"],
        user=config["SQL_USER_NAME"],
        password=config["SQL_PASSWORD"],
        database="DB"
    )


def create_table(cursor: MySQLCursor) -> None:
    create_table = f"""
    CREATE TABLE IF NOT EXISTS {TABLE_NAME} (
        bio_id VARCHAR(7),
        last_name VARCHAR(25),
        first_name VARCHAR(20),
        state CHAR(2),
        party VARCHAR(35),
        congress INT,
        chamber CHAR(1),
        PRIMARY KEY (bio_id, congress, chamber)
    );
    """
    cursor.execute(create_table)


def insert_records(records: List[Record],
                   cursor: MySQLCursor,
                   db: MySQLConnection) -> None:
    insert = f"""
    INSERT IGNORE INTO {TABLE_NAME}
    (bio_id, last_name, first_name, state, party, congress, chamber)
    VALUES (%s, %s, %s, %s, %s, %s, %s)
    """
    values = [
        (
            record.bio_id,
            record.last_name,
            record.first_name,
            record.state,
            record.party,
            record.congress,
            record.chamber
        )
        for record in records
    ]
    cursor.executemany(insert, values)
    db.commit()


def main(records: List[Record]) -> None:
    try:
        config = get_config()
        db = connect_to_db(config)
        cursor = db.cursor()
        create_table(cursor)
        print("Started inserting")
        insert_records(records, cursor, db)
        print("Finished inserting")
    except Exception as e:
        print(f"\n{__file__}\nError | {e}")
    finally:
        if cursor:
            cursor.close()
        if db:
            db.close()

File: data/bills/insertRecords.js
const sql = require("mysql2/promise");
const pool = require("../db");


async function createTable(connection)
{
    const query = `
        CREATE TABLE IF NOT EXISTS Bills (
            bill_id VARCHAR(15) PRIMARY KEY,
            congress TINYINT,
            type VARCHAR(7),
            bill_num VARCHAR(5),
            vote CHAR(1),
            h_vote_num VARCHAR(5),
            h_year SMALLINT,
            s_vote_num VARCHAR(5),
            s_session TINYINT
        );
    `;
    try {
        await connection.execute(query);
    } catch (error) {
        console.error(`
            ${__filename}
            Error in createTable() | ${error}
        `);
    }
}


async function insertRecords(connection, records)
{
    const query = `
        INSERT IGNORE INTO Bills
        (bill_id, congress, type, bill_num, vote, h_vote_num, h_year, s_vote_num, s_session)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?);
    `
    for (const record of records) {
        try {
            await connection.execute(query, [
                record.bill_id,
                record.congress,
                record.type,
                record.bill_num,
                record.vote,
                record.h_vote_num,
                record.h_year,
                record.s_vote_num,
                record.s_session
            ]);
        } catch (error) {
            console.error(`
                ${__filename}
                Error in insertRecords() | ${error}
            `);
        }
    }
}


async function main(records)
{
    const connection = await pool.getConnection();
    await createTable(connection);
    await insertRecords(connection, records);
    connection.release();
}


module.exports = main;
